File: 341/341_version1.py
class NestedIterator(object):

    def __init__(self, nestedList):
        """
        Initialize your data structure here.
        :type nestedList: List[NestedInteger]
        """
        self.array = self.flattenArray(nestedList)
        self.index = 0
        

    def next(self):
        """
        :rtype: int
        """
        if self.hasNext():
            value = self.array[self.index]
            self.index+=1
            return value

    def hasNext(self):
        """
        :rtype: bool
        """
        return self.index < len(self.array)
    
    def flattenArray(self,array):
        result = []
        for i in array:
            if i.isInteger():
                result.append(i.getInteger())
            else:
                result.extend(self.flattenArray(i.getList()))
        return result

File: 341/341_version1.java
public class NestedIterator implements Iterator<Integer> {
    private int index;
    private List<Integer> array;
    public NestedIterator(List<NestedInteger> nestedList) {
        array = flattenArray(nestedList);
        index = 0;
    }

    @Override
    public Integer next() {
        if(hasNext()){
            int value = array.get(index);
            index+=1;
            return value;
        }
        return -1;
    }

    @Override
    public boolean hasNext() {
        return index < array.size();
        
    }
    
    private List<Integer> flattenArray(List<NestedInteger> nestedList){
        List<Integer> array = new ArrayList<>();
        for(NestedInteger n:nestedList){
            if(n.isInteger()==true) array.add(n.getInteger());
            else{
                for(int i:flattenArray(n.getList())) array.add(i);
            }
        }
        return array;
    }
}
